client/client.go

```diff
@@ -229,22 +242,31 @@ func (cli *Client) negotiateAPIVersionPing(p types.Ping) {
 	}
 
 	// if the client is not initialized with a version, start with the latest supported version
-	if cli.version == "" {
-		cli.version = api.DefaultVersion
+	if cli.ClientVersion() == "" {
```

Comment:
Perhaps we should just take the lock for this whole function, to make it a more atomic operation, instead of using the accessors (currently it's potentially possible to have a different result here than on the lines below)